deps/status/src/checkLockfilesUpToDate.ts

```diff
@@ -0,0 +1,380 @@
+
+      const wantedLockfilePromise = readWantedLockfile(workspaceDir, { ignoreIncompatible: false })
+      if (wantedLockfileStats.mtime.valueOf() > packagesList.lastValidatedTimestamp) {
+        const virtualStoreDir = path.join(workspaceDir, 'node_modules', '.pnpm')
+        const currentLockfile = await readCurrentLockfile(virtualStoreDir, { ignoreIncompatible: false })
+        const wantedLockfile = (await wantedLockfilePromise) ?? throwLockfileNotFound(workspaceDir)
+        assertLockfilesEqual(currentLockfile, wantedLockfile, workspaceDir)
+      }
+      readWantedLockfileAndDir = async () => ({
+        wantedLockfile: (await wantedLockfilePromise) ?? throwLockfileNotFound(workspaceDir),
+        wantedLockfileDir: workspaceDir,
+      })
+    } else {
+      readWantedLockfileAndDir = async wantedLockfileDir => {
+        const wantedLockfilePromise = readWantedLockfile(wantedLockfileDir, { ignoreIncompatible: false })
+        const [
+          wantedLockfileStats,
+        ] = await Promise.all([
+          statIfExists(path.join(wantedLockfileDir, WANTED_LOCKFILE)),
+        ])
+
+        if (!wantedLockfileStats) return throwLockfileNotFound(wantedLockfileDir)
+        if (wantedLockfileStats.mtime.valueOf() > packagesList.lastValidatedTimestamp) {
+          const virtualStoreDir = path.join(wantedLockfileDir, 'node_modules', '.pnpm')
+          const currentLockfile = await readCurrentLockfile(virtualStoreDir, { ignoreIncompatible: false })
+          const wantedLockfile = (await wantedLockfilePromise) ?? throwLockfileNotFound(wantedLockfileDir)
+          assertLockfilesEqual(currentLockfile, wantedLockfile, wantedLockfileDir)
+        }
+
+        return {
+          wantedLockfile: (await wantedLockfilePromise) ?? throwLockfileNotFound(wantedLockfileDir),
+          wantedLockfileDir,
+        }
+      }
+    }
+
+    type GetProjectId = (project: Pick<Project, 'rootDir'>) => ProjectId
+    const getProjectId: GetProjectId = sharedWorkspaceLockfile
+      ? project => getLockfileImporterId(workspaceDir, project.rootDir)
+      : () => '.' as ProjectId
+
+    const getWorkspacePackages = once(arrayOfWorkspacePackagesToMap.bind(null, allProjects))
+    const getManifestsByDir = once(() => getWorkspacePackagesByDirectory(getWorkspacePackages()))
+
+    await Promise.all(modifiedProjects.map(async ({ project }) => {
+      const { wantedLockfile, wantedLockfileDir } = await readWantedLockfileAndDir(project.rootDir)
+
+      await assertWantedLockfileUpToDate({
+        autoInstallPeers,
+        config: opts,
+        excludeLinksFromLockfile,
+        linkWorkspacePackages,
+        getManifestsByDir,
+        getWorkspacePackages,
+        projectDir: project.rootDir,
+        projectId: getProjectId(project),
+        projectManifest: project.manifest,
+        rootDir: workspaceDir,
+        rootManifestOptions,
+        wantedLockfile,
+        wantedLockfileDir,
+      })
```

Comment:
I’d divide the arguments to 2. The first a context that is the same for every call and the second one with project, wantedLockfile and wantedLockfileDir.

Reply:
https://github.com/pnpm/pnpm/pull/8645/commits/85ba93ad7e7573a814f8a213d264ad15f327834d